pkg/upgrade/upgrade.go

```diff
@@ -652,3 +678,701 @@ func cleanupDeprecatedKueueVAPB(ctx context.Context, cli client.Client) error {
+func calculateResourceLimitsFromSizes(containerSizes []ContainerSize) (map[string]string, error) {
+	var maxMemory, minMemory, maxCpu, minCpu *resource.Quantity
+
+	for _, size := range containerSizes {
+		// Process memory values
+		memMax, memMin, err := calculateMemoryLimits(size.Resources.Requests.Memory, size.Resources.Limits.Memory)
+		if err != nil {
+			return nil, err
+		}
+		maxMemory = selectMaxQuantity(maxMemory, memMax)
+		minMemory = selectMinQuantity(minMemory, memMin)
+
+		// Process CPU values
+		cpuMax, cpuMin, err := calculateCpuLimits(size.Resources.Requests.Cpu, size.Resources.Limits.Cpu)
+		if err != nil {
+			return nil, err
+		}
+		maxCpu = selectMaxQuantity(maxCpu, cpuMax)
+		minCpu = selectMinQuantity(minCpu, cpuMin)
+	}
+
+	// Apply defaults if no values found
+	result := make(map[string]string)
+
+	if minMemory == nil {
+		defaultMem := resource.MustParse(defaultMinMemory)
+		minMemory = &defaultMem
+	}
+	if minCpu == nil {
+		defaultCpu := resource.MustParse(defaultMinCpu)
+		minCpu = &defaultCpu
+	}
+
+	result["minMemory"] = minMemory.String()
+	result["minCpu"] = minCpu.String()
+
+	if maxMemory != nil {
+		result["maxMemory"] = maxMemory.String()
+	}
+	if maxCpu != nil {
+		result["maxCpu"] = maxCpu.String()
+	}
+
+	return result, nil
+}
+
+// calculateResourceLimits processes request and limit values and returns max/min quantities.
+func calculateResourceLimits(request, limit, resourceType string) (*resource.Quantity, *resource.Quantity, error) {
+	if request == "" && limit == "" {
+		return nil, nil, nil
+	}
+
+	var requestQty, limitQty *resource.Quantity
+	var err error
+
+	if request != "" {
+		requestQty, err = parseQuantity(request)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to parse %s request %q: %w", resourceType, request, err)
+		}
+	}
+
+	if limit != "" {
+		limitQty, err = parseQuantity(limit)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to parse %s limit %q: %w", resourceType, limit, err)
+		}
+	}
+
+	// Get the larger of request and limit for max, smaller for min
+	switch {
+	case requestQty != nil && limitQty != nil:
+		if requestQty.Cmp(*limitQty) > 0 {
+			return requestQty, limitQty, nil
+		}
+		return limitQty, requestQty, nil
+	case requestQty != nil:
+		return requestQty, requestQty, nil
+	case limitQty != nil:
+		return limitQty, limitQty, nil
+	default:
+		return nil, nil, nil
+	}
+}
+
+// calculateMemoryLimits processes memory request and limit values and returns max/min quantities.
+func calculateMemoryLimits(requestMemory, limitMemory string) (*resource.Quantity, *resource.Quantity, error) {
+	return calculateResourceLimits(requestMemory, limitMemory, "memory")
+}
+
+// calculateCpuLimits processes CPU request and limit values and returns max/min quantities.
+func calculateCpuLimits(requestCpu, limitCpu string) (*resource.Quantity, *resource.Quantity, error) {
+	return calculateResourceLimits(requestCpu, limitCpu, "CPU")
+}
```

Comment:
not sure about the reason of those two functions, since in fact this is just a wrapper on `calculateResourceLimits` with a different string, so `calculateResourceLimits` can be used directly